File: Traditional-Important-Algorithms/exercises_py/02_memory_models/mm02_raii.py
# I AM NOT DONE

"""
Exercise: RAII (Resource Acquisition Is Initialization)

RAII is a programming idiom where resource lifetime is tied to object lifetime.
Resources are acquired in constructor and released in destructor.

Your task: Implement RAII wrappers for different resources.
"""

# Simulated file system
_file_system = {}


def _init_fs():
    """Initialize simulated file system"""
    global _file_system
    _file_system = {}


def _write_file(path, content):
    """Write to simulated file"""
    _file_system[path] = content


def _file_exists(path):
    """Check if file exists"""
    return path in _file_system


def _delete_file(path):
    """Delete file from simulated file system"""
    if path in _file_system:
        del _file_system[path]


class FileHandle:
    """RAII file handle"""

    def __init__(self, path):
        # TODO: Implement RAII file handle
        # Create file in simulated file system
        pass

    def write(self, content):
        """Write to file if not closed"""
        # TODO: Write to file if not closed
        pass

    def close(self):
        """Mark as closed"""
        # TODO: Mark as closed (cleanup will happen in __del__)
        pass

    def is_closed(self):
        """Check if file is closed"""
        return self.closed

    def __del__(self):
        """Destructor - automatic cleanup"""
        # TODO: Implement automatic cleanup
        # Delete file from simulated file system
        pass

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()
        return False


# Lock implementation
_lock_state = False


class Lock:
    """RAII lock"""

    def __init__(self):
        # TODO: Acquire lock using RAII
        # Check if lock is available, acquire it
        # Raise exception if lock is not available
        pass

    def __del__(self):
        """Destructor - automatically release lock"""
        # TODO: Automatically release lock
        pass

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        return False


def _is_locked():
    """Check if lock is held"""
    return _lock_state


import unittest


class TestRAII(unittest.TestCase):
    def setUp(self):
        """Setup for each test"""
        _init_fs()
        global _lock_state
        _lock_state = False

    def test_file_handle_raii(self):
        _init_fs()

        # Use in block scope
        file = FileHandle("test.txt")
        file.write("hello")
        self.assertTrue(_file_exists("test.txt"))
        del file

        # File should be deleted when file is destroyed
        self.assertFalse(_file_exists("test.txt"))

    def test_explicit_close(self):
        _init_fs()

        file = FileHandle("test.txt")
        file.close()
        self.assertTrue(file.is_closed())

        result = file.write("data")
        self.assertIsNone(result)

    def test_lock_raii(self):
        # Use in block scope
        lock = Lock()
        self.assertTrue(_is_locked())

        # Can't acquire twice
        try:
            lock2 = Lock()
            self.fail("Should not be able to acquire lock twice")
        except:
            pass

        del lock

        # Lock should be released
        self.assertFalse(_is_locked())

        # Can acquire again
        lock3 = Lock()
        self.assertTrue(_is_locked())

    def test_file_context_manager(self):
        _init_fs()

        with FileHandle("test.txt") as f:
            f.write("hello")
            self.assertTrue(_file_exists("test.txt"))
            self.assertFalse(f.is_closed())

        # File should be closed after context
        self.assertTrue(f.is_closed())

    def test_lock_context_manager(self):
        with Lock():
            self.assertTrue(_is_locked())

        self.assertFalse(_is_locked())

    def test_multiple_files(self):
        _init_fs()

        f1 = FileHandle("file1.txt")
        f2 = FileHandle("file2.txt")

        f1.write("content1")
        f2.write("content2")

        self.assertTrue(_file_exists("file1.txt"))
        self.assertTrue(_file_exists("file2.txt"))

        del f1

        self.assertFalse(_file_exists("file1.txt"))
        self.assertTrue(_file_exists("file2.txt"))

    def test_file_write_after_close(self):
        _init_fs()

        f = FileHandle("test.txt")
        f.write("first")

        f.close()

        result = f.write("second")
        self.assertIsNone(result)

    def test_nested_locks_fail(self):
        lock1 = Lock()

        with self.assertRaises(Exception):
            lock2 = Lock()

    def test_sequential_locks(self):
        lock1 = Lock()
        self.assertTrue(_is_locked())

        del lock1
        self.assertFalse(_is_locked())

        lock2 = Lock()
        self.assertTrue(_is_locked())

        del lock2
        self.assertFalse(_is_locked())

    def test_file_survives_until_deleted(self):
        _init_fs()

        f = FileHandle("test.txt")
        f.write("content")

        self.assertTrue(_file_exists("test.txt"))

        # File persists while handle exists
        for _ in range(10):
            self.assertTrue(_file_exists("test.txt"))

        del f

        self.assertFalse(_file_exists("test.txt"))

    def test_lock_released_on_exception(self):
        try:
            lock = Lock()
            self.assertTrue(_is_locked())
            raise ValueError("test exception")
        except ValueError:
            pass

        # Lock should still be released after exception
        # (Python's garbage collection handles this)
        # Force cleanup
        import gc
        gc.collect()

        # After GC, lock should be released
        # Note: This behavior depends on Python's GC
        self.assertFalse(_is_locked())


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/02_memory_models/mm02_raii.rs
// mm02_raii.rs
//
// RAII (Resource Acquisition Is Initialization) is a programming idiom
// where resource lifetime is tied to object lifetime. Resources are
// acquired in constructor and released in destructor.
//
// Your task: Implement RAII wrappers for different resources.

// I AM NOT DONE

use std::collections::HashMap;

// Simulated file system
static mut FILE_SYSTEM: Option<HashMap<String, String>> = None;

fn init_fs() {
    unsafe {
        FILE_SYSTEM = Some(HashMap::new());
    }
}

fn write_file(path: &str, content: &str) {
    unsafe {
        if let Some(fs) = &mut FILE_SYSTEM {
            fs.insert(path.to_string(), content.to_string());
        }
    }
}

fn file_exists(path: &str) -> bool {
    unsafe {
        FILE_SYSTEM.as_ref()
            .map(|fs| fs.contains_key(path))
            .unwrap_or(false)
    }
}

fn delete_file(path: &str) {
    unsafe {
        if let Some(fs) = &mut FILE_SYSTEM {
            fs.remove(path);
        }
    }
}

pub struct FileHandle {
    path: String,
    closed: bool,
}

impl FileHandle {
    pub fn open(path: &str) -> Self {
        // TODO: Implement RAII file handle
        // Create file in simulated file system
        todo!()
    }

    pub fn write(&mut self, content: &str) -> Result<(), &'static str> {
        // TODO: Write to file if not closed
        todo!()
    }

    pub fn close(&mut self) {
        // TODO: Mark as closed (cleanup will happen in Drop)
        todo!()
    }

    pub fn is_closed(&self) -> bool {
        self.closed
    }
}

impl Drop for FileHandle {
    fn drop(&mut self) {
        // TODO: Implement automatic cleanup
        // Delete file from simulated file system
        todo!()
    }
}

// Lock implementation
static mut LOCK_STATE: bool = false;

pub struct Lock {
    acquired: bool,
}

impl Lock {
    pub fn acquire() -> Result<Self, &'static str> {
        // TODO: Acquire lock using RAII
        // Check if lock is available, acquire it
        todo!()
    }
}

impl Drop for Lock {
    fn drop(&mut self) {
        // TODO: Automatically release lock
        todo!()
    }
}

fn is_locked() -> bool {
    unsafe { LOCK_STATE }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_file_handle_raii() {
        init_fs();

        {
            let mut file = FileHandle::open("test.txt");
            file.write("hello").unwrap();
            assert!(file_exists("test.txt"));
        } // file should be deleted when dropped

        assert!(!file_exists("test.txt"));
    }

    #[test]
    fn test_explicit_close() {
        init_fs();

        let mut file = FileHandle::open("test.txt");
        file.close();
        assert!(file.is_closed());
        assert!(file.write("data").is_err());
    }

    #[test]
    fn test_lock_raii() {
        {
            let _lock = Lock::acquire().unwrap();
            assert!(is_locked());
            assert!(Lock::acquire().is_err()); // Can't acquire twice
        } // lock released

        assert!(!is_locked());
        assert!(Lock::acquire().is_ok()); // Can acquire again
    }
}
